### PR TITLE
Update: MultiputUpload should use upload host from preflight response

### PR DESCRIPTION
**Pull request by wenboyu2**
- create `BaseUpload` class to share preflight request and error handling
 - move `PlainUpload` to `src/api/uploads`
 - `MultiputUpload` now uses the upload host from the preflight response

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @wenboyu2 has signed the CLA. Thanks for the pull request!

**Comment by boxcla** [comment]
# Webpack Bundle Size Github Bot

| Bundle        | Diff | Percentage Diff | Before | After |
| ------------- |:----:|:---------------:|:------:|:-----:|
| Total | 0 B | 0.00% | 0 B | 0 B |

Compared against previously merged PR #319

**Powered by [webpack-bundle-size-github-bot](https://github.com/wenboyu2/webpack-bundle-size-github-bot/)**

File: src/api/uploads/PlainUpload.js [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+     *
+     * @param {Object} data - Upload success data
+     * @return {void}
+     */
+    uploadSuccessHandler = ({ entries }: { entries: BoxItem[] }): void => {
+        if (this.isDestroyed()) {
+            return;
+        }
+
+        if (typeof this.successCallback === 'function') {
```
Comment (priyajeet) [inline src/api/uploads/PlainUpload.js]: check needed when defaulted to noop?

File: src/api/uploads/PlainUpload.js [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+     *
+     * @param {Object} event - Progress event
+     * @return {void}
+     */
+    uploadProgressHandler = (event: ProgressEvent): void => {
+        if (this.isDestroyed()) {
+            return;
+        }
+
+        if (typeof this.progressCallback === 'function') {
```
Comment (priyajeet) [inline src/api/uploads/PlainUpload.js]: same comment

File: src/api/uploads/PlainUpload.js [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+     * Cancels upload of a file.
+     *
+     * @return {void}
+     */
+    cancel() {
+        if (this.isDestroyed()) {
+            return;
+        }
+
+        if (this.xhr && typeof this.xhr.abort === 'function') {
```
Comment (priyajeet) [inline src/api/uploads/PlainUpload.js]: I think this will always be true

File: src/api/uploads/MultiputUpload.js [hunk 4]
```diff
@@ -143,22 +138,41 @@ class MultiputUpload extends BaseMultiput {
...
+     * @param {Object} [response.data]
+     * @return {string}
+     */
+    getBaseUploadUrlFromPreflightResponse = ({ data }: { data: { upload_url?: string } }) => {
+        if (!data || !data.upload_url) {
+            return this.getBaseUploadUrl();
+        }
+
+        const splitUrl = data.upload_url.split('/');
+        this.uploadHost = `${splitUrl[0]}//${splitUrl[2]}`;
```
Comment (priyajeet) [inline src/api/uploads/MultiputUpload.js]: two // ?
Can you put an example as comments

File: src/api/uploads/BaseUpload.js [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+     * @return {void}
+     */
+    preflightErrorHandler = (error: any): void => {
+        if (this.isDestroyed()) {
+            return;
+        }
+
+        this.fileName = this.file ? this.file.name : '';
+
+        // TODO(tonyjin): Normalize error object and clean up error handling
```
Comment (priyajeet) [inline src/api/uploads/BaseUpload.js]: @tonyjin 😡 👇

File: src/api/uploads/PlainUpload.js (hunk 1 above)
Comment (wenboyu2) [reply src/api/uploads/PlainUpload.js]: I just copy pasted PlainUpload.js into another directory, maybe @tonyjin can answer comments related to PlainUpload?

File: src/api/uploads/MultiputUpload.js (hunk 4 above)
Comment (wenboyu2) [reply src/api/uploads/MultiputUpload.js]: `splitUrl[0]` is the protocol part (e.g., https:), `splitUrl[2]` is hostname (e.g., www.box.com)

Sure I'll add comments

**Pull request merged**